velox/functions/lib/Re2Functions.cpp

```diff
@@ -970,89 +970,266 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> re2ExtractSignatures() {
+    cursor++;
+  }
+
+  return os.str();
+}
+
+// An iterator to iterate through a pattern string. It will handle escaping
+// automatically.
+class PatternStringIterator {
+ public:
+  PatternStringIterator(StringView pattern, std::optional<char> escapeChar)
+      : pattern_(pattern),
+        escapeChar_(escapeChar),
+        lastIndex_{pattern_.size() - 1} {}
+
+  // Advance the cursor to next char, escape char is automatically handled.
+  bool next() {
+    if (currentIndex_ == lastIndex_) {
+      return false;
+    }
+
+    isPreviousWildcard_ =
+        (charKind_ == CharKind::kSingleCharWildcard ||
+         charKind_ == CharKind::kAnyCharsWildcard);
+
+    currentIndex_++;
+    auto currentChar = current();
+    if (currentChar == escapeChar_) {
+      // Escape char should be followed by another char.
+      VELOX_USER_CHECK_LT(
+          currentIndex_,
+          lastIndex_,
+          "Escape character must be followed by '%', '_' or the escape character itself: {}, escape {}",
+          pattern_,
+          escapeChar_.value())
+
+      currentIndex_++;
+      currentChar = current();
+      // The char follows escapeChar can only be one of (%, _, escapeChar).
+      if (currentChar == escapeChar_ || currentChar == '_' ||
+          currentChar == '%') {
+        charKind_ = CharKind::kNormal;
+      } else {
+        VELOX_USER_FAIL(
+            "Escape character must be followed by '%', '_' or the escape character itself: {}, escape {}",
+            pattern_,
+            escapeChar_.value())
+      }
+    } else if (currentChar == '_') {
+      charKind_ = CharKind::kSingleCharWildcard;
+    } else if (currentChar == '%') {
+      charKind_ = CharKind::kAnyCharsWildcard;
+    } else {
+      charKind_ = CharKind::kNormal;
+    }
+
+    return true;
+  }
+
+  // Current index of the cursor.
+  char currentIndex() const {
+    return currentIndex_;
+  }
+
+  bool isAnyCharsWildcard() const {
+    return charKind_ == CharKind::kAnyCharsWildcard;
+  }
+
+  bool isSingleCharWildcard() const {
+    return charKind_ == CharKind::kSingleCharWildcard;
+  }
+
+  bool isWildcard() {
+    return isAnyCharsWildcard() || isSingleCharWildcard();
+  }
+
+  bool isPreviousWildcard() {
+    return isPreviousWildcard_;
+  }
+
+ private:
+  // Represents the state of current cursor/char.
+  enum class CharKind {
+    // Wildcard char: %.
+    // NOTE: If escape char is set as '\', for pattern '\%%', the first '%' is
+    // not a wildcard, just a literal '%', the second '%' is a wildcard.
+    kAnyCharsWildcard,
+    // Wildcard char: _.
+    // NOTE: If escape char is set as '\', for pattern '\__', the first '_' is
+    // not a wildcard, just a literal '_', the second '_' is a wildcard.
+    kSingleCharWildcard,
+    // Chars that are not escape char & not wildcard char.
+    kNormal
+  };
+
+  // Char at current cursor.
+  char current() const {
+    return pattern_.data()[currentIndex_];
+  }
+
+  const StringView pattern_;
```

Comment:
I feel we'll have to take another pass, maybe in a follow-up PR, to replace StringView with std::string_view. StringView is really meant to store strings in vectors and is not a replacement for light-weight std::string_view. For example, for short strings we end up copying strings when passing StringView by value and for these strings it is not safe to store a pointer to data().

Reply:
Agreed, I will submit a dedicated a PR for StringView change.